### PR TITLE
Testing against multiple solutions and multiple input/output-pairs

### PR DESCRIPTION
**Pull request by Arxcis**
## TLDR;
A lot of stuff is changing, but the repo will still be familiar to you :+1: 
image

## Why?

1. Make it possible for everyone to submit their own input, without worrying about getting conflicts with others.

2.  It is helpful to test against multiple distinct input/output-pairs, to make sure our solutions are robust - does not make assumptions about a specific input/output-files being provided.
```
arxcis.input
arxcis.output
preng.input
preng.output
```

## Suggested change

Change the signature of the language-tests to take a list of `SOLUTION_FILES` and a list of `IO_FILES` as input:

*Example declaration in `languages/go.sh`*
```sh
#!/usr/bin/env bash
set -euo pipefail

# Usage:      ../../lang/go.sh  "SOLUTION_FILES"   "IO_FILES"
#
# Example:    ../../lang/go.sh  "solutions/*.go"   "io/*"
# Expands to: ../../lang/go.sh   solutions/main.go  io/alice.input io/alice.output io/bob.input io/bob.output
#
SOLUTION_FILES=$1   # Expand $1 into list
IO_FILES=$2         # Expand $2 into list
```

*Example usage in `days/day-05/test.sh`*
```sh
echo ""
echo "--- Day 5: Boarding pass ---"

$D/../../lang/go.sh     "$D/solutions/*.go"   "$D/io/*"
$D/../../lang/c.sh      "$D/solutions/*.c"    "$D/io/*"
$D/../../lang/sml.sh    "$D/solutions/*.sml"  "$D/io/*"
$D/../../lang/python.sh "$D/solutions/*.py"   "$D/io/*"
$D/../../lang/deno.sh   "$D/solutions/*.ts"   "$D/io/*"

echo ""
```

*Example running `./days/day-05/test.sh`*
```
--- Day 5: Boarding pass ---
cat INPUT | go       day05.stektpotet.go     2ms        ✅
cat INPUT | go       tholok97.go             2ms        ✅
cat INPUT | gcc      day05.c                 2ms        ✅
cat INPUT | polyc    day05.sml               4ms        ✅
cat INPUT | python3  day05.klyve.py          15ms       ✅
cat INPUT | python3  day05.preng.py          11ms       ✅
cat INPUT | python3  day05.stektpotet.py     11ms       ✅
cat INPUT | python3  one-liner.stektpotet.py 12ms       ✅
cat INPUT | deno     day05.ts                15ms       ✅
```

### CODE REVIEW

**Comment by tholok97** [comment]
I'll have a look after work @Arxcis :)

File: days/day-00-example/test.sh [hunk 1]
```diff
@@ -5,16 +5,16 @@ D=$(dirname $(realpath $0))
...
-$D/../../languages/ruby.sh   $D/input.txt $D/output.txt $D/solutions/example.rb
-$D/../../languages/java.sh   $D/input.txt $D/output.txt $D/solutions Example
+$D/../../languages/rust.sh   $D/io/alice.* $D/solutions/example.rs
+$D/../../languages/go.sh     $D/io/alice.* $D/solutions/example.go
+$D/../../languages/c.sh      $D/io/alice.* $D/solutions/example.c
+$D/../../languages/cpp.sh    $D/io/alice.* $D/solutions/example.cpp
+$D/../../languages/sml.sh    $D/io/alice.* $D/solutions/example.sml
+$D/../../languages/bash.sh   $D/io/alice.* $D/solutions/example.bash
+$D/../../languages/php.sh    $D/io/alice.* $D/solutions/example.php
+$D/../../languages/python.sh $D/solutions/example.py $D/io/*
```
Comment (Arxcis) [inline days/day-00-example/test.sh]: :thinking: 

Should we do the same globbing for all solutions for a language within a given day?
```suggestion
$D/../../languages/python.sh $D/solutions/*.py $D/io/*
```

Scope creep...  Ignore! I am leaving it for a potential later PR, unless someone can see a trivial way to implement this in `bash`, because I only see complex ways :thinking: 
```suggestion
$D/../../languages/python.sh --solutions $D/solutions/*.py --io $D/io/*
```

Comment (Arxcis) [reply days/day-00-example/test.sh:15]: ....although, this would actually enable us to have a `test.sh` at root-level, which runs all the tests for a single language, across all days, which would be pretty neat :thinking:

Comment (Arxcis) [reply days/day-00-example/test.sh]: **UPDATE** I actually found a trivial solution to the problem presented above :tada:  See e07e3bd for details :eyes:

Comment (Arxcis) [reply days/day-00-example/test.sh]: But now we are saying goodbye to sorting by performance...

Comment (Arxcis) [reply days/day-00-example/test.sh]: We might as well go as far as doing: `../../test-day.sh "solutions/*" "io/*"` at this point, and just don't sort at all.

Comment (Arxcis) [reply days/day-00-example/test.sh]: ....unless someone has a trivial way to sort a list by a column in bash... :thinking:

Comment (Arxcis) [reply days/day-00-example/test.sh]: Resolving this conversation, by reverting e07e3bd, because it introduces complexity and indirection, which I am not comfortable to introduce at this point in time.

**Comment by tholok97** [comment]
@Arxcis sorry I got hung up doing aoc with a friend over discord, and then I hit my head against the wall with lisp for a while.. I don't know when I'll have time to take a proper look at this. I have a suggestion, but I don't feel comfortable giving it until I've tried to look for myself how it will work in practice.

(Btw behold my creation:

image

A precursor to how I want to do day 1 in Lisp)

**Comment by Arxcis** [comment]
@tholok97 No worries. I have the feeling, I am going to need a few days to settle on something myself. This PR is not currently building in the CI, so you can hold off reviewing, until I have at least figured that one out.

We can do the review/discussion/grooming in IRL if you like :wink: The good thing with this PR, is that we will eventually have something to look at while discussing .

**Comment by Arxcis** [comment]
You know what. I got a breakthrough. The tests are passing, and I am happy with how the new architecture works. I am comfortable with setting this to "Ready for review" now, as I am going to bed :bed: :sleeping:

**Comment by Arxcis** [comment]
I got a really good review from @tholok97  IRL :smile:  Thank you very much!  :pray: I will take all the things into consideration when making a decision on what to do tonight.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -4,34 +4,53 @@
...
+```sh
+./days/day03/test.sh
+# or
+make test.day03
+# or
+make docker.day03
+```
+
+7. When you are happy with local testing, make a Pull Request to the `main` branch.
+8. One of the maintainers will merge PR's, at the of each day.
```
Comment (Stektpotet) [inline CONTRIBUTING.md:33]: One of the maintainers will merge PR's, at the ................. of each day.

File: scripts/make-folders.sh [hunk 3]
```diff
@@ -1,35 +1,35 @@
...
-    cat > "./$DAY/solutions/.keep" << KEEP
-KEEP
+    touch "./$DAY/solutions/.keep"
+
+    mkdir -p "./$DAY/io/"
+    touch "./$DAY/io/.keep"
 
     cat > "./$DAY/README.md" << README
 # $DAY
 ## --- Part 1 ---
```
Comment (Stektpotet) [inline scripts/make-folders.sh:17]: I'm pretty sure most of the readme fiiles we have so far have these two combined.

It could be changed to this:

`# ---  $DAY: Title ---`

Not  that that really matters in the template, as it's going to be written over at some point again...
¯\\\_(ツ)\_/¯

**Review by Stektpotet — COMMENTED** [review]
From what I saw, only a couple of minor things need to/can be "fixed" here.
As a whole, nice merge 👍

File: CONTRIBUTING.md (hunk 2 above)
Comment (Arxcis) [reply CONTRIBUTING.md:33]: :eyes: Well spotted. Since I already merged, I will commit a fix for this directly :+1:

File: scripts/make-folders.sh (hunk 3 above)
Comment (Arxcis) [reply scripts/make-folders.sh:17]: I think it matters, even if it is a minor thing, because the template is supposed to help the developer. It is a way of documenting things, and communicating intent, so that any contributor is capable of contributing README.md's. I will fix in a commit directly. :+1:

**Comment by Arxcis** [comment]
Thank you for reviewing @Stektpotet :pray:
